packages/synthetics-sdk-broken-links/src/broken_links.ts

```diff
@@ -12,24 +12,193 @@
+  // navigate to each link - LOOP:
+  //          each call to `checkLinks(...)` will return a `SyntheticLinkResult`
+  //          Object added to an array of `followed_links`
+
+  // returned a SyntheticResult with `options`, `followed_links` &
+  // runtimeMetadata
   return {} as SyntheticResult;
 }
+
+async function checkLink(
+  page: Page,
+  link: LinkIntermediate,
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions
+): Promise<BrokenLinksResultV1_SyntheticLinkResult> {
+  // START - to resolve warnings while under development
+  page;
+  link;
+  options;
+  // END - to resolve warnings  while under development
+
+  // PSEUDOCODE
+
+  // determine expected_status_code—determined here rather than `navigate(...)`
+  // since it will be used later when creating
+  // `BrokenLinksResultV1_SyntheticLinkResult` Object
+
+  // call `navigate(...)`
+  /**
+   * const {
+   *    response: responseOrError,
+   *    passed,
+   *    testing_only,
+   *    link_start_time,
+   *    link_end_time,
+   * } = await navigate(page, link, expected_status_code, options);
+   */
+
+  // Error handling based on `responseOrError` & passed:
+  //    if      `responseOrError` is type `Error` set `error_type` and
+  //            `error_message` in `SyntheticLinkResult` accordingly
+  //    else if `passed === false`. Set error `SYNTHETIC_INCORRECT_STATUS_CODE`
+
+  // return `SynheticLinkResult` with all calculated information
+  return {} as BrokenLinksResultV1_SyntheticLinkResult;
+}
+
+/**
+ * Navigates to a target URL with retries and timeout handling.
+ *
+ * @param page - The Puppeteer Page instance.
+ * @param link - The LinkIntermediate containing the target URL.
+ * @param expected_status_code - The expected HTTP status code.
+ * @param options - The options for navigation and retries.
+ * @returns Information about navigation attempt:
+ *   - `responseOrError`: HTTP response or error if navigation fails, or null.
+ *   - `passed`: Boolean indicating if navigation passed per status code.
+ *   - `retriesRemaining`: Remaining retries after attempt. (for testing)
+ *   - `link_start_time`: Start time of navigation attempt.
+ *   - `link_end_time`: End time of navigation attempt.
+ */
+export async function navigate(
+  page: Page,
+  link: LinkIntermediate,
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions,
+  expected_status_code: ResponseStatusCode = {
+    status_class: ResponseStatusCode_StatusClass.STATUS_CLASS_2XX,
+  }
+): Promise<NavigateResponse> {
+  let fetch_link_output = {} as CommonResponseProps;
+  let retriesRemaining = options.max_retries! + 1;
+  // use link_specific timeout if set, else use options.link_timeout_millis
+  const per_link_timeout_millis =
+    options.per_link_options[link.target_url]?.link_timeout_millis ||
+    options.link_timeout_millis!;
+
+  if (shouldGoToBlankPage(page.url(), link.target_url)) {
+    await page.goto('about:blank');
+  }
+
+  let passed = false;
```

Comment:
Seeing `passed` as a retry condition and in the result looks weird -- but from our chat, this is actually intentional. Can you add that context? Especially any non-obvious product considerations

Reply:
done.